### PR TITLE
[Merged by Bors] - feat(Algebra/BigOperators/Group): add `sum_tsub_distrib`

### PR DESCRIPTION
**Pull request by Command-Master**
Upstreamed from https://github.com/b-mehta/ExponentialRamsey/blob/7e17b629a915a082869f494c8afa56a3e1c7a88d/ExponentialRamsey/Prereq/Mathlib/Algebra/BigOperators/Ring.lean#L17

Co-authored-by: b-mehta <bhavikmehta8@gmail.com>

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/BigOperators/Ring.lean [hunk 1]
```diff
@@ -22,6 +22,14 @@ variable {ι ι' α β γ : Type*} {κ : ι → Type*} {s s₁ s₂ : Finset ι}
 
 namespace Finset
 
+theorem sum_tsub_distrib {α β : Type*} [AddCommMonoid β] [PartialOrder β] [ExistsAddOfLE β]
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Ring.lean:25]:
```suggestion
theorem sum_tsub_distrib [AddCommMonoid β] [PartialOrder β] [ExistsAddOfLE β]
```

File: Mathlib/Algebra/BigOperators/Ring.lean [hunk 2]
```diff
@@ -22,6 +22,14 @@ variable {ι ι' α β γ : Type*} {κ : ι → Type*} {s s₁ s₂ : Finset ι}
...
+    [CovariantClass β β (· + ·) (· ≤ ·)] [ContravariantClass β β (· + ·) (· ≤ ·)] [Sub β]
+    [OrderedSub β] (s : Finset α) {f g : α → β} (hfg : ∀ x ∈ s, g x ≤ f x) :
+    ∑ x in s, (f x - g x) = ∑ x in s, f x - ∑ x in s, g x :=
+  eq_tsub_of_add_eq <| by
+    rw [← Finset.sum_add_distrib];
+    exact Finset.sum_congr rfl fun x hx => tsub_add_cancel_of_le <| hfg _ hx
```
Comment (b-mehta) [inline Mathlib/Algebra/BigOperators/Ring.lean:31]:
```suggestion
    rw [← sum_add_distrib]
    exact sum_congr rfl fun x hx => tsub_add_cancel_of_le <| hfg _ hx
```
I think this works?

**Review by b-mehta — APPROVED** [review]
LGTM but I wrote this so I won't merge.

File: Mathlib/Algebra/BigOperators/Ring.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Algebra/BigOperators/Ring.lean:25]: Actually, can you switch to using `ι` for the index type and `α` for the summand type to match the rest of the file?

File: Mathlib/Algebra/BigOperators/Ring.lean [hunk 3]
```diff
@@ -22,6 +22,14 @@ variable {ι ι' α β γ : Type*} {κ : ι → Type*} {s s₁ s₂ : Finset ι}
 
 namespace Finset
 
+theorem sum_tsub_distrib [AddCommMonoid α] [PartialOrder α] [ExistsAddOfLE α]
```
Comment (eric-wieser) [inline Mathlib/Algebra/BigOperators/Ring.lean:25]: This apparently can be placed in `Mathlib.Algebra.BigOperators.Group.Finset`, which seems more appropriate.

Can you add the multiset version too?

File: Mathlib/Algebra/BigOperators/Group/Finset.lean [hunk 4]
```diff
@@ -1478,6 +1478,11 @@ theorem sum_range_tsub [AddCommMonoid α] [PartialOrder α] [Sub α] [OrderedSub
     have h₂ : f 0 ≤ f n := h (Nat.zero_le _)
     rw [tsub_add_eq_add_tsub h₂, add_tsub_cancel_of_le h₁]
 
+theorem sum_tsub_distrib [AddCommMonoid α] [PartialOrder α] [ExistsAddOfLE α]
+    [CovariantClass α α (· + ·) (· ≤ ·)] [ContravariantClass α α (· + ·) (· ≤ ·)] [Sub α]
+    [OrderedSub α] (s : Finset ι) {f g : ι → α} (hfg : ∀ x ∈ s, g x ≤ f x) :
+      ∑ x ∈ s, (f x - g x) = ∑ x ∈ s, f x - ∑ x ∈ s, g x := sum_map_tsub _ hfg
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/BigOperators/Group/Finset.lean:1484]:
```suggestion
    ∑ x ∈ s, (f x - g x) = ∑ x ∈ s, f x - ∑ x ∈ s, g x := sum_map_tsub _ hfg
```

**Review by jcommelin — COMMENTED** [review]
bors d+

**Comment by urkud** [comment]
Thanks! :tada:
bors merge

**Pull request closed without merging**